lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3544,13 +3567,23 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     // Note: we relax the restriction if the type we're casting to is a
     // non-final class because it's possible that we might have a subclass
     // that conforms to the protocol.
+    //
+    //  Also, relax the restriction when toType is a stdlib collection
+    //  (Array, Set, Dictionary) that could have conditional conformances
+    //  because of custom casting mechanism implemented for those types e.g.
+    //
+    //  func encodable(_ value: Encodable) {
+    //    _ = value as! [String : Encodable]
+    //  }
+    //
     if (fromExistential && !toExistential) {
       if (auto NTD = toType->getAnyNominal()) {
         if (!toType->is<ClassType>() || NTD->isFinal()) {
           auto protocolDecl =
               dyn_cast_or_null<ProtocolDecl>(fromType->getAnyNominal());
-          if (protocolDecl &&
-              !conformsToProtocol(toType, protocolDecl, dc)) {
+          if (protocolDecl && !conformsToProtocol(toType, protocolDecl, dc) &&
+              !(toType->isStdlibCollectionType() &&
+                getStdlibModule(dc)->lookupConformance(toType, protocolDecl))) {
```

Comment:
You'll want to lookup the conformance from the current module, which you can get with `dc->getParentModule()`. Also IMO it may be worth factoring out this logic into a `couldDynamicallyConformTo` static function so we can take advantage of some early returns.

Reply:
@hamishknight I think this is the only thing left right? I'm adjusting right now and this will be ready for review :)

Reply:
Yup, everything else looks good to me!

Reply:
@hamishknight let me know if is that what you have in mind, not really sure if I got the idea here